quickwit-search/src/export/leaf.rs

```diff
@@ -0,0 +1,127 @@
+use crate::OutputFormat;
+use crate::SearchError;
+use quickwit_proto::LeafExportResult;
+use quickwit_storage::Storage;
+use std::sync::Arc;
+use tantivy::schema::Type;
+use tantivy::{query::Query, ReloadPolicy};
+use tokio_stream::wrappers::ReceiverStream;
+
+/// `leaf` step of export.
+// Note: we return a stream of a result with a tonic::Status error
+// to be compatible with the stream coming from the grpc client.
+// It would be better to have a SearchError but we need then
+// to process stream in grpc_adapater.rs to change SearchError
+// to tonic::Status as tonic::Status is required by the stream result
+// signature defined by proto generated code.
+pub async fn leaf_export(
+    query: Box<dyn Query>,
+    fast_field_collector_builder: FastFieldCollectorBuilder,
+    split_ids: Vec<String>,
+    storage: Arc<dyn Storage>,
+    output_format: OutputFormat,
+) -> ReceiverStream<Result<LeafExportResult, tonic::Status>> {
+    let (result_sender, result_receiver) = tokio::sync::mpsc::channel(10);
+    for split_id in split_ids {
+        let split_storage: Arc<dyn Storage> =
+            quickwit_storage::add_prefix_to_storage(storage.clone(), split_id.clone());
+        let collector_for_split = fast_field_collector_builder.clone();
+        let query_clone = query.box_clone();
+        let result_sender_clone = result_sender.clone();
+        tokio::spawn(async move {
+            let leaf_split_result = leaf_export_single_split(
+                query_clone,
+                collector_for_split,
+                split_storage,
+                output_format,
+            )
+            .await
+            .map_err(SearchError::convert_to_tonic_status);
+            result_sender_clone
+                .send(leaf_split_result)
+                .await
+                .map_err(|_| {
+                    SearchError::InternalError(format!(
+                        "Unable to send leaf export result for split `{}`",
+                        split_id
+                    ))
+                })?;
+            Result::<(), SearchError>::Ok(())
+        });
+    }
+    ReceiverStream::new(result_receiver)
+}
+
+/// Apply a leaf search on a single split.
+async fn leaf_export_single_split(
+    query: Box<dyn Query>,
+    fast_field_collector_builder: FastFieldCollectorBuilder,
+    storage: Arc<dyn Storage>,
+    output_format: OutputFormat,
+) -> crate::Result<LeafExportResult> {
+    let index = open_index(storage).await?;
+    let reader = index
+        .reader_builder()
+        .num_searchers(1)
+        .reload_policy(ReloadPolicy::Manual)
+        .try_into()?;
+    let searcher = reader.searcher();
+    warmup(
+        &*searcher,
+        query.as_ref(),
+        fast_field_collector_builder.fast_field_to_warm(),
+    )
+    .await?;
+    let mut buffer = Vec::new();
+    match fast_field_collector_builder.value_type() {
+        Type::I64 => {
+            let fast_field_collector = fast_field_collector_builder.build_i64();
```

Comment:
I suspect it feels a bit bad to have the same code copy paste.
I think this is perfectly fine here.

There are only 2 cases. Trying to do something smart would be overengineering.